packages/filler/src/services/ContractInteractionService.ts

```diff
@@ -399,4 +504,272 @@ export class ContractInteractionService {
+		const inputs = order.inputs
+
+		for (const output of outputs) {
+			const tokenAddress = output.token
+			const amount = output.amount
+			const decimals = await this.getTokenDecimals(tokenAddress, order.destChain)
+			const price = await this.getTokenPrice(tokenAddress)
+			const tokenUsdValue = (amount / BigInt(10 ** decimals)) * price
+			outputUsdValue = outputUsdValue + tokenUsdValue
+		}
+
+		for (const input of inputs) {
+			const tokenAddress = input.token
+			const amount = input.amount
+			const decimals = await this.getTokenDecimals(tokenAddress, order.sourceChain)
+			const price = await this.getTokenPrice(tokenAddress)
+			const tokenUsdValue = (amount / BigInt(10 ** decimals)) * price
+			inputUsdValue = inputUsdValue + tokenUsdValue
+		}
+
+		return { outputUsdValue, inputUsdValue }
+	}
+
+	async getTokenPrice(tokenAddress: string): Promise<bigint> {
+		const usdValue = await fetchTokenUsdPriceOnchain(tokenAddress)
+		return usdValue
+	}
+
+	async getFillerBalanceUSD(
+		order: Order,
+		chain: string,
+	): Promise<{
+		nativeTokenBalance: bigint
+		daiBalance: bigint
+		usdtBalance: bigint
+		usdcBalance: bigint
+		totalBalanceUsd: bigint
+	}> {
+		const fillerWalletAddress = privateKeyToAddress(this.privateKey)
+		const destClient = this.clientManager.getPublicClient(chain)
+
+		const nativeTokenBalance = await destClient.getBalance({ address: fillerWalletAddress })
+		const nativeTokenPriceUsd = await this.getNativeTokenPriceUsd(order)
+		const nativeTokenUsdValue = (nativeTokenBalance / BigInt(10 ** 18)) * nativeTokenPriceUsd
+
+		// DAI balance
+		const daiAddress = this.configService.getDaiAsset(chain)
+		const daiBalance = await destClient.readContract({
+			abi: ERC20_ABI,
+			address: daiAddress,
+			functionName: "balanceOf",
+			args: [fillerWalletAddress],
+		})
+		const daiDecimals = await this.getTokenDecimals(daiAddress, chain)
+		const daiBalanceUsd = daiBalance / BigInt(10 ** daiDecimals)
+
+		// USDT Balance
+		const usdtAddress = this.configService.getUsdtAsset(chain)
+		const usdtBalance = await destClient.readContract({
+			abi: ERC20_ABI,
+			address: usdtAddress,
+			functionName: "balanceOf",
+			args: [fillerWalletAddress],
+		})
+		const usdtDecimals = await this.getTokenDecimals(usdtAddress, chain)
+		const usdtBalanceUsd = usdtBalance / BigInt(10 ** usdtDecimals)
+
+		// USDC Balance
+		const usdcAddress = this.configService.getUsdcAsset(chain)
+		const usdcBalance = await destClient.readContract({
+			abi: ERC20_ABI,
+			address: usdcAddress,
+			functionName: "balanceOf",
+			args: [fillerWalletAddress],
+		})
+		const usdcDecimals = await this.getTokenDecimals(usdcAddress, chain)
+		const usdcBalanceUsd = usdcBalance / BigInt(10 ** usdcDecimals)
+
+		const totalBalanceUsd = nativeTokenUsdValue + daiBalanceUsd + usdtBalanceUsd + usdcBalanceUsd
+
+		return { nativeTokenBalance, daiBalance, usdtBalance, usdcBalance, totalBalanceUsd }
+	}
+
+	async calculateSwapOperations(
```

Comment:
this should be encapsulated within the SwapStrategy. Unless its used by other strategies

Reply:
Okay, I will move it there.